File: uis/client/web2/src/services/auth_service.ts
import { AppConfig } from '../config/app_config';
import { User } from '../models/user';
import { apiService } from './api_service';
import { userService } from './user_service';

export class AuthService {
  // Login
  async login(email: string, password: string): Promise<User> {
    const response = await apiService.post(`${AppConfig.authEndpoint}/auth/login`, {
      email,
      password,
    });

    if (response.status === 200) {
      // Defensive: response.data is unknown, so cast after checking
      const data = response.data as Record<string, unknown>;

      if (typeof data === 'object' && data !== null) {
        // Store tokens
        if ('access_token' in data && typeof data.access_token === 'string') {
          localStorage.setItem(AppConfig.accessTokenKey, data.access_token);
        }
        if ('refresh_token' in data && typeof data.refresh_token === 'string') {
          localStorage.setItem(AppConfig.refreshTokenKey, data.refresh_token);
        }

        // Store keycloak_id if available in token
        let keycloakId: string | null = null;
        if ('access_token' in data && typeof data.access_token === 'string') {
          try {
            // Decode JWT to get keycloak_id
            const tokenParts = data.access_token.split('.');
            if (tokenParts.length === 3) {
              let payloadBase64 = tokenParts[1];
              while (payloadBase64.length % 4) {
                payloadBase64 += '=';
              }
              const payload = JSON.parse(atob(payloadBase64));
              keycloakId = payload.sub || payload.keycloak_id || null;
              if (keycloakId) {
                localStorage.setItem('keycloak_id', keycloakId);
              }
            }
          } catch (e) {
            console.error('Failed to decode token:', e);
          }
        }

        // Try to fetch user profile - but don't fail login if it doesn't exist yet
        if (keycloakId) {
          try {
            // Try /user endpoint first
            const user = await userService.getUserProfile(keycloakId);
            localStorage.setItem(AppConfig.userDataKey, JSON.stringify(user));
            return User.fromJson(user as unknown as Record<string, unknown>);
          } catch (profileError) {
            console.warn('User profile not found, trying /me endpoint:', profileError);
            // try {
            //   const meResponse = await apiService.get(`${AppConfig.authEndpoint}/me`);
            //   if (
            //     meResponse.status === 200 &&
            //     typeof meResponse.data === 'object' &&
            //     meResponse.data !== null &&
            //     'data' in meResponse.data &&
            //     typeof (meResponse.data as { data?: unknown }).data === 'object' &&
            //     (meResponse.data as { data?: unknown }).data !== null
            //   ) {
            //     const user = User.fromJson((meResponse.data as { data: Record<string, unknown> }).data);
            //     localStorage.setItem(AppConfig.userDataKey, JSON.stringify(user.toJson()));
            //     return user;
            //   }
            // } catch (meError) {
            //   console.warn('Failed to fetch user from /me endpoint:', meError);
            // }
            // If both fail, return a minimal user object - login should still succeed
            console.warn('User profile not available, creating minimal user object');
                const minimalUser = User.fromJson({
                  id: keycloakId,
                  email: email,
                  firstName: '',
                  lastName: '',
                  status: 'active',
                  isVerified: false,
                  createdAt: new Date().toISOString(),
                });
                localStorage.setItem('keycloak_id', keycloakId);
                localStorage.setItem(AppConfig.userDataKey, JSON.stringify(minimalUser.toJson()));
                return minimalUser;
          }
        }

        // If we can't decode keycloak_id, still allow login but with minimal user
        console.warn('Could not extract keycloak_id from token, proceeding with minimal user');
        const minimalUser = User.fromJson({
          id: email, // Use email as fallback ID
          email: email,
          firstName: '',
          lastName: '',
          status: 'active',
          isVerified: false,
          createdAt: new Date().toISOString(),
        });
        localStorage.setItem(AppConfig.userDataKey, JSON.stringify(minimalUser.toJson()));
        return minimalUser;
      }
    }
    throw new Error('Login failed');
  }

  // Register - Just store data (all setup happens at customer creation)
  async register(params: {
    email: string;
    password: string;
    firstName: string;
    lastName: string;
    phoneNumber: string;
    uin?: string;
  }): Promise<{ email: string }> {
    // Store onboarding data for later use (will be sent to orchestrator at PIN creation)
    localStorage.setItem('onboarding_data', JSON.stringify({
      email: params.email,
      password: params.password,
      firstName: params.firstName,
      lastName: params.lastName,
      phoneNumber: params.phoneNumber,
      uin: params.uin,
    }));

    return { email: params.email };
  }

  // Get current user from API or storage fallback
  async getCurrentUser(): Promise<User | null> {
    try {
      const response = await apiService.get(`${AppConfig.authEndpoint}/me`);
      if (response.status === 200) {
        const data = response.data as { data: Record<string, unknown> };
        const user = User.fromJson(data.data);
        // Extract and store keycloak_id if available in user data
        const keycloakId = (user as { keycloakId?: string; keycloak_id?: string }).keycloakId || (user as { keycloak_id?: string }).keycloak_id || (data.data && typeof data.data === 'object' ? (data.data as { keycloak_id?: string; keycloakId?: string }).keycloak_id ?? (data.data as { keycloakId?: string }).keycloakId : undefined);
        if (keycloakId) {
          localStorage.setItem('keycloak_id', keycloakId);
        }
        // Store user data
        localStorage.setItem(AppConfig.userDataKey, JSON.stringify(user.toJson()));
        return user;
      }
      return null;
    } catch {
      // Fallback to stored user
      const userData = localStorage.getItem(AppConfig.userDataKey);
      if (userData) {
        const user = User.fromJson(JSON.parse(userData));
        // Try to extract keycloak_id from stored user
        const storedUser = JSON.parse(userData);
        const keycloakId = storedUser.keycloakId || storedUser.keycloak_id;
        if (keycloakId && !localStorage.getItem('keycloak_id')) {
          localStorage.setItem('keycloak_id', keycloakId);
        }
        return user;
      }
      return null;
    }
  }



// Add this new method
getUser(): User | null {
  return this.getStoredUser();
}


  // Logout
  async logout(): Promise<void> {
    try {
      await apiService.post(`${AppConfig.authEndpoint}/logout`);
    } catch {
      // Continue with local logout even if API fails
    } finally {
      localStorage.removeItem(AppConfig.accessTokenKey);
      localStorage.removeItem(AppConfig.refreshTokenKey);
      localStorage.removeItem(AppConfig.userDataKey);
    }
  }

  // Check if user is authenticated
  isAuthenticated(): boolean {
    const token = localStorage.getItem(AppConfig.accessTokenKey);
    return !!token;
  }

  // Get stored user
  getStoredUser(): User | null {
    const userData = localStorage.getItem(AppConfig.userDataKey);
    if (userData) return User.fromJson(JSON.parse(userData));
    return null;
  }

  // Forgot password
  async forgotPassword(email: string): Promise<{ keycloak_id: string; otp_expires_at: string }> {
    const response = await apiService.post(`${AppConfig.authEndpoint}/auth/forgot-password`, { email });
    return response.data as { keycloak_id: string; otp_expires_at: string };
  }

  // Reset password
  async resetPassword(params: {
    keycloakId: string;
    otpCode: string;
    newPassword: string;
    confirmPassword: string;
  }): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/auth/reset-password`, {
      keycloak_id: params.keycloakId,
      otp_code: params.otpCode,
      new_password: params.newPassword,
      confirm_password: params.confirmPassword,
    });
  }

  // Change password
  async changePassword(currentPassword: string, newPassword: string, confirmPassword: string): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/auth/change-password`, {
      current_password: currentPassword,
      new_password: newPassword,
      confirm_password: confirmPassword,
    });
  }

  // create PIN
  async createPin(currentPin: string, newPin: string): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/create-pin`, {
      current_pin: currentPin,
      new_pin: newPin,
    });
  }

  // Verify email
  async verifyEmail(token: string): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/verify-email`, { token });
  }

  // Resend verification email
  async resendVerificationEmail(): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/resend-verification`);
  }

  // Verify OTP
  async verifyOtp(otp: string): Promise<void> {
    const response = await apiService.post(`${AppConfig.authEndpoint}/verify-otp`, { otp });
    if (response.status === 200) {
      let data: unknown;
      if (response.data && typeof response.data === 'object' && 'data' in response.data) {
        data = (response.data as { data: Record<string, unknown> }).data;
      } else {
        data = response.data;
      }
      if (typeof data === 'object' && data !== null && 'access_token' in data) {
        const tokenData = data as { access_token: string; refresh_token?: string };
        localStorage.setItem(AppConfig.accessTokenKey, tokenData.access_token);
        if (tokenData.refresh_token) {
          localStorage.setItem(AppConfig.refreshTokenKey, tokenData.refresh_token);
        }
        // Decode JWT to get keycloak_id
        try {
          const tokenParts = tokenData.access_token.split('.');
          if (tokenParts.length === 3) {
            let payloadBase64 = tokenParts[1];
            while (payloadBase64.length % 4) {
              payloadBase64 += '=';
            }
            const payload = JSON.parse(atob(payloadBase64));
            const keycloakId = payload.sub || payload.keycloak_id;
            if (keycloakId) {
              localStorage.setItem('keycloak_id', keycloakId);
            }
          }
        } catch (e) {
          console.error('Failed to decode token:', e);
        }
      }
    }
  }

  // Resend OTP
  async resendOtp(): Promise<void> {
    await apiService.post(`${AppConfig.authEndpoint}/resend-otp`);
  }
}

export const authService = new AuthService();

File: uis/client/mobile_app/lib/services/auth_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../config/app_config.dart';
import '../models/user.dart';
import 'api_service.dart';
import 'user_service.dart';
import 'local_storage_service.dart';
import 'dart:convert';

class AuthService {
  final ApiService _apiService = ApiService();
  final UserService _userService = UserService();
  final FlutterSecureStorage _storage = const FlutterSecureStorage(
    aOptions: AndroidOptions(
      encryptedSharedPreferences: true,
      resetOnError: true,
    ),
  );

  // Login
  Future<User> login(String email, String password) async {
    final response = await _apiService.post(
      '${AppConfig.authEndpoint}/auth/login',
      data: {
        'email': email,
        'password': password,
      },
    );

    if (response.statusCode == 200) {
      final data = response.data;

      // Store tokens
      if (data['access_token'] != null) {
      await _storage.write(
        key: AppConfig.accessTokenKey,
        value: data['access_token'],
      );
      }
      if (data['refresh_token'] != null) {
      await _storage.write(
        key: AppConfig.refreshTokenKey,
        value: data['refresh_token'],
      );
      }

      // Get keycloak_id from response data (matching web app)
      String? keycloakId;
      // First try to get keycloak_id directly from response data
      keycloakId = data['keycloak_id']?.toString() ?? data['data']?['keycloak_id']?.toString();
      
      // If not in response, try to decode from token (fallback)
      if (keycloakId == null && data['access_token'] != null) {
        try {
          final tokenParts = (data['access_token'] as String).split('.');
          if (tokenParts.length == 3) {
            // Add padding if needed for base64 decoding
            String payloadBase64 = tokenParts[1];
            while (payloadBase64.length % 4 != 0) {
              payloadBase64 += '=';
            }
            final payloadBytes = base64Decode(payloadBase64);
            final payload = jsonDecode(utf8.decode(payloadBytes)) as Map<String, dynamic>;
            keycloakId = payload['sub'] ?? payload['keycloak_id'];
          }
        } catch (e) {
          print('Failed to decode token: $e');
        }
      }
      
      if (keycloakId != null) {
        // Store in LocalStorageService so API service can read it
        await LocalStorageService.setString('keycloak_id', keycloakId);
      }

      // Try to fetch user profile - but don't fail login if it doesn't exist yet
      if (keycloakId != null) {
        try {
          // Try getUserProfile first
          final user = await _userService.getUserProfile(keycloakId);
          await _storage.write(
            key: AppConfig.userDataKey,
            value: jsonEncode(user.toJson()),
          );
          return user;
        } catch (profileError) {
          print('User profile not found, trying alternative methods: $profileError');
          try {
            // Try getCurrentUser as fallback
            final user = await _userService.getCurrentUser(keycloakId);
            if (user != null) {
              await _storage.write(
                key: AppConfig.userDataKey,
                value: jsonEncode(user.toJson()),
              );
              return user;
            }
          } catch (currentUserError) {
            print('Failed to fetch user from getCurrentUser: $currentUserError');
          }
          
          // If both fail, return a minimal user object - login should still succeed
          print('User profile not available, creating minimal user object');
          final minimalUser = User.fromJson({
            'id': keycloakId,
            'email': email,
            'first_name': '',
            'last_name': '',
            'status': 'active',
            'is_verified': false,
            'created_at': DateTime.now().toIso8601String(),
          });
          await _storage.write(
            key: AppConfig.userDataKey,
            value: jsonEncode(minimalUser.toJson()),
          );
          return minimalUser;
        }
      }

      // If we can't decode keycloak_id, still allow login but with minimal user
      print('Could not extract keycloak_id from token, proceeding with minimal user');
      final minimalUser = User.fromJson({
        'id': email, // Use email as fallback ID
        'email': email,
        'first_name': '',
        'last_name': '',
        'status': 'active',
        'is_verified': false,
        'created_at': DateTime.now().toIso8601String(),
      });
      await _storage.write(
        key: AppConfig.userDataKey,
        value: jsonEncode(minimalUser.toJson()),
      );
      return minimalUser;
    } else {
      throw Exception('Login failed');
    }
  }

  // Register
  Future<User> register({
    required String email,
    required String password,
    required String firstName,
    required String lastName,
    required String phoneNumber,
  }) async {
    // Setup auth profile
    await _apiService.post(
      '${AppConfig.authEndpoint}/auth',
      data: {
        'email': email,
        'user_role': "user",
      },
    );

    // Setup password
    await _apiService.post(
      '${AppConfig.authEndpoint}/setup-password',
      data: {
        "password": password,
        "confirm_password": password,
      },
    );

    // Create user through user service
    final user = await _userService.createUser(
      email: email,
      password: password,
      firstName: firstName,
      lastName: lastName,
      phoneNumber: phoneNumber,
    );

    return user;
  }

  // Get current user (delegates to UserService)
  Future<User?> getCurrentUser(String keycloakId) async {
    return await _userService.getCurrentUser(keycloakId);
  }

  // Get stored user (delegates to UserService)
  Future<User?> getStoredUser() async {
    return await _userService.getStoredUser();
  }

  // Logout
  Future<void> logout() async {
    try {
      await _apiService.post('${AppConfig.authEndpoint}/logout');
    } catch (e) {
      // Continue with local logout even if API call fails
    } finally {
      await _storage.deleteAll();
    }
  }

  // Check if user is authenticated
  Future<bool> isAuthenticated() async {
    final token = await _storage.read(key: AppConfig.accessTokenKey);
    return token != null && token.isNotEmpty;
  }

  // Forgot password
  Future<Map<String, dynamic>> forgotPassword(String email) async {
    final response = await _apiService.post(
      '${AppConfig.authEndpoint}/auth/forgot-password',
      data: {'email': email},
    );

    final data = response.data as Map<String, dynamic>;
    final keycloakId = data['keycloak_id']?.toString();
    if (keycloakId != null && keycloakId.isNotEmpty) {
      await LocalStorageService.setString('password_reset_keycloak_id', keycloakId);
    }

    return data;
  }

  // Reset password
  Future<void> resetPassword({
    required String keycloakId,
    required String otpCode,
    required String newPassword,
    required String confirmPassword,
  }) async {
    await _apiService.post(
      '${AppConfig.authEndpoint}/auth/reset-password',
      data: {
        'keycloak_id': keycloakId,
        'otp_code': otpCode,
        'new_password': newPassword,
        'confirm_password': confirmPassword,
      },
    );

    await LocalStorageService.remove('password_reset_keycloak_id');
  }

  // Change password
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
    required String confirmPassword,
  }) async {
    await _apiService.post(
      '${AppConfig.authEndpoint}/auth/change-password',
      data: {
        'current_password': currentPassword,
        'new_password': newPassword,
        'confirm_password': confirmPassword,
      },
    );
  }

  // Verify email
  Future<void> verifyEmail(String token) async {
    await _apiService.post(
      '${AppConfig.authEndpoint}/verify-email',
      data: {'token': token},
    );
  }

  // Resend verification email
  Future<void> resendVerificationEmail() async {
    await _apiService.post('${AppConfig.authEndpoint}/resend-verification');
  }
}
